lib/src/media/stream/client.dart

```diff
@@ -5,51 +5,112 @@ import 'package:flutter_webrtc/flutter_webrtc.dart';
 import '../../domain/web_rtc/web_rtc_client/web_rtc_client.dart';
 import '../../gen/proto/stream/v1/stream.pbgrpc.dart';
 
-class StreamClient {
+class StreamManager {
+  final Map<String, MediaStream> _streams = {};
+  final Map<String, StreamClient> _clients = {};
   final WebRtcClientChannel _channel;
   final StreamServiceClient _client;
-  // ignore: close_sinks
-  final StreamController<MediaStream> _streamController = StreamController<MediaStream>.broadcast();
-  MediaStream? _stream;
   StreamSubscription? _errorHandler;
 
-  StreamClient(this._channel) : _client = StreamServiceClient(_channel) {
+  static final Finalizer<StreamSubscription?> _finalizer = Finalizer((p0) {
+    p0?.cancel();
+  });
+
+  StreamManager(this._channel) : _client = StreamServiceClient(_channel) {
+    _finalizer.attach(this, _errorHandler);
     _channel.rtcPeerConnection.onAddStream = (MediaStream stream) {
-      _stream = stream;
-      _streamController.add(stream);
+      addStream(stream);
     };
 
     _channel.rtcPeerConnection.onConnectionState = (state) {
       if (state == RTCPeerConnectionState.RTCPeerConnectionStateFailed ||
           state == RTCPeerConnectionState.RTCPeerConnectionStateDisconnected) {
         _errorHandler = Stream.periodic(const Duration(seconds: 1)).listen((_) {
-          _streamController.addError(Exception('PeerConnection error'));
+          for (final client in _clients.values) {
+            client._streamController.addError(Exception('PeerConnection error'));
+          }
         });
       }
     };
   }
 
-  String _getValidSDPTrackName(String name) {
+  void addStream(MediaStream stream) {
+    _streams[stream.id] = stream;
+    if (_clients.containsKey(stream.id)) {
+      _clients[stream.id]!._internalStreamController.add(stream);
+    }
+  }
+
+  String getValidSDPTrackName(String name) {
     return name.replaceAll(':', '+');
   }
 
-  Stream<MediaStream> getStream(String name) {
-    if (_stream == null) {
-      final future = add(name);
-      future.onError((error, stackTrace) => _streamController.addError(error ?? Exception('Could not add stream named $name')));
+  StreamClient getStreamClient(String name) {
+    final sanitizedName = getValidSDPTrackName(name);
+    if (_clients.containsKey(sanitizedName)) {
+      return _clients[sanitizedName]!;
+    }
+    final client = StreamClient(name, remove);
+    _clients[sanitizedName] = client;
+
+    if (_streams.containsKey(sanitizedName)) {
+      Future.delayed(const Duration(milliseconds: 100), () {
```

Comment:
Removed it here but left it below. Basically we need the delay because broadcast streams don't buffer events, so we need to send the event AFTER the stream has been returned (so that it can actually be listened to). this was the fastest way of doing it, rather than creating my own stream (which we might want to look into later)

Reply:
i see, sounds good.